File: 00-Reference/20250718-beginner/ts/solution.test.ts
import { Calculator } from './solution';
import testCases from './test-cases.json';

describe('Calculator', () => {
  let calc: Calculator;

  beforeEach(() => {
    calc = new Calculator();
  });

  testCases.forEach((testCase) => {
    it(testCase.name, () => {
      const [a, b] = testCase.input;
      
      if (testCase.expected === "error") {
        expect(() => {
          switch (testCase.operation) {
            case 'add':
              calc.add(a, b);
              break;
            case 'subtract':
              calc.subtract(a, b);
              break;
            case 'multiply':
              calc.multiply(a, b);
              break;
            case 'divide':
              calc.divide(a, b);
              break;
          }
        }).toThrow();
      } else {
        let result: number;
        switch (testCase.operation) {
          case 'add':
            result = calc.add(a, b);
            break;
          case 'subtract':
            result = calc.subtract(a, b);
            break;
          case 'multiply':
            result = calc.multiply(a, b);
            break;
          case 'divide':
            result = calc.divide(a, b);
            break;
          default:
            throw new Error(`Unknown operation: ${testCase.operation}`);
        }
        expect(result).toBe(testCase.expected);
      }
    });
  });
});

File: 00-Reference/20250718-beginner/go/solution_test.go
package main

import (
	"testing"
)

func TestCalculator(t *testing.T) {
	calc := NewCalculator()

	// Test case 1: Basic Addition
	result := calc.Add(5, 3)
	expected := 8
	if result != expected {
		t.Errorf("Add(5, 3) = %d; expected %d", result, expected)
	}

	// Test case 2: Basic Subtraction
	result = calc.Subtract(10, 4)
	expected = 6
	if result != expected {
		t.Errorf("Subtract(10, 4) = %d; expected %d", result, expected)
	}

	// Test case 3: Basic Multiplication
	result = calc.Multiply(7, 2)
	expected = 14
	if result != expected {
		t.Errorf("Multiply(7, 2) = %d; expected %d", result, expected)
	}

	// Test case 4: Basic Division
	result, err := calc.Divide(15, 3)
	expected = 5
	if err != nil {
		t.Errorf("Divide(15, 3) returned error: %v", err)
	}
	if result != expected {
		t.Errorf("Divide(15, 3) = %d; expected %d", result, expected)
	}

	// Test case 5: Division with Remainder
	result, err = calc.Divide(10, 3)
	expected = 3
	if err != nil {
		t.Errorf("Divide(10, 3) returned error: %v", err)
	}
	if result != expected {
		t.Errorf("Divide(10, 3) = %d; expected %d", result, expected)
	}

	// Test case 6: Division by Zero
	_, err = calc.Divide(10, 0)
	if err == nil {
		t.Error("Divide(10, 0) should return error")
	}

	// Test case 7: Negative Numbers
	result = calc.Add(-5, 3)
	expected = -2
	if result != expected {
		t.Errorf("Add(-5, 3) = %d; expected %d", result, expected)
	}

	// Test case 8: Zero Operations
	result = calc.Multiply(0, 5)
	expected = 0
	if result != expected {
		t.Errorf("Multiply(0, 5) = %d; expected %d", result, expected)
	}

	// Test case 9: Large Numbers
	result = calc.Add(1000, 2000)
	expected = 3000
	if result != expected {
		t.Errorf("Add(1000, 2000) = %d; expected %d", result, expected)
	}

	// Test case 10: Subtraction Result Negative
	result = calc.Subtract(3, 5)
	expected = -2
	if result != expected {
		t.Errorf("Subtract(3, 5) = %d; expected %d", result, expected)
	}
}